lib/pbio/drv/usb/usb_ev3.c

```diff
@@ -906,122 +901,148 @@ void pbdrv_usb_schedule_status_update(const uint8_t *status_msg) {
     pbio_os_request_poll();
 }
 
-static pbio_os_process_t pbdrv_usb_ev3_process;
+// True if we are expected to send a response to a recent incoming message.
+static bool usb_send_response;
 
-static pbio_error_t pbdrv_usb_ev3_process_thread(pbio_os_state_t *state, void *context) {
-    static pbio_os_timer_t delay_timer;
-    static pbio_os_timer_t tx_timeout_timer;
-    static pbio_os_timer_t keepalive_timer;
-    static bool was_transmitting = false;
-    static bool is_transmitting = false;
+/**
+ * Non-blocking poll handler to process pending incoming messages.
+ */
+static void pbdrv_usb_handle_data_in(void) {
+    if (!usb_rx_is_ready) {
+        return;
+    }
 
-    PBIO_OS_ASYNC_BEGIN(state);
+    // This barrier prevents *subsequent* memory reads from being
+    // speculatively moved *earlier*, outside the if statement
+    // (which is technically allowed by the as-if rule).
+    pbdrv_compiler_memory_barrier();
 
-    for (;;) {
-        if (pbdrv_usb_config == 0) {
-            pbdrv_usb_is_events_subscribed = false;
-            // Resend status flags when host subscribes
-            pbdrv_usb_status_data_pending = true;
+    uint32_t usb_rx_sz = PBDRV_UNCACHED(cppi_descriptors[CPPI_DESC_RX].word0).pktLength;
+    pbio_pybricks_error_t result;
+
+    // Skip empty commands.
+    if (usb_rx_sz) {
+        pbdrv_cache_prepare_after_dma(ep1_rx_buf, sizeof(ep1_rx_buf));
+
+        switch (ep1_rx_buf[0]) {
+            case PBIO_PYBRICKS_OUT_EP_MSG_SUBSCRIBE:
+                pbdrv_usb_is_events_subscribed = ep1_rx_buf[1];
+                ep1_tx_response_buf[0] = PBIO_PYBRICKS_IN_EP_MSG_RESPONSE;
+                pbio_set_uint32_le(&ep1_tx_response_buf[1], PBIO_PYBRICKS_ERROR_OK);
+                usb_send_response = true;
+                // Request to resend status flags now that the host has subscribed.
+                // Our response will take priority, status follows right after.
+                pbdrv_usb_status_data_pending = true;
+                break;
+            case PBIO_PYBRICKS_OUT_EP_MSG_COMMAND:
+                if (!pbdrv_usb_receive_handler) {
+                    break;
+                }
+                result = pbdrv_usb_receive_handler(ep1_rx_buf + 1, usb_rx_sz - 1);
+                ep1_tx_response_buf[0] = PBIO_PYBRICKS_IN_EP_MSG_RESPONSE;
+                pbio_set_uint32_le(&ep1_tx_response_buf[1], result);
+                usb_send_response = true;
+                break;
```

Comment:
The original version would immediately send the response here. In the revision, we only schedule it here and transmit it from the process when we get round to it.